db_test.go

```diff
@@ -1601,6 +1603,396 @@ func TestCorrectNumTombstones(t *testing.T) {
+		{
+			mint: []int64{0, 14, 5},
+			maxt: []int64{9, 22, 17},
+			blockSeries: [][]Series{
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{0, 0}, sample{1, 0}, sample{2, 0}, sample{4, 0},
+						sample{5, 0}, sample{8, 0}, sample{9, 0},
+					}),
+				},
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{14, 59}, sample{15, 59}, sample{17, 59}, sample{20, 59},
+						sample{21, 59}, sample{22, 59},
+					}),
+				},
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{5, 99}, sample{6, 99}, sample{7, 99}, sample{8, 99},
+						sample{9, 99}, sample{10, 99}, sample{13, 99}, sample{15, 99},
+						sample{16, 99}, sample{17, 99},
+					}),
+				},
+			},
+			exp: map[string][]sample{`{a="b"}`: {
+				sample{0, 0}, sample{1, 0}, sample{2, 0}, sample{4, 0},
+				sample{5, 99}, sample{6, 99}, sample{7, 99}, sample{8, 99},
+				sample{9, 99}, sample{10, 99}, sample{13, 99}, sample{14, 59},
+				sample{15, 59}, sample{16, 99}, sample{17, 59}, sample{20, 59},
+				sample{21, 59}, sample{22, 59},
+			}},
+		},
+		// Case 5
+		// |-------------------------------------|
+		//            |------------|
+		//      |-------------------------|
+		{
+			mint: []int64{0, 7, 3},
+			maxt: []int64{22, 11, 17},
+			blockSeries: [][]Series{
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{0, 0}, sample{1, 0}, sample{2, 0}, sample{4, 0},
+						sample{5, 0}, sample{8, 0}, sample{9, 0}, sample{10, 0},
+						sample{13, 0}, sample{15, 0}, sample{16, 0}, sample{17, 0},
+						sample{20, 0}, sample{22, 0},
+					}),
+				},
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{7, 59}, sample{8, 59}, sample{9, 59}, sample{10, 59},
+						sample{11, 59},
+					}),
+				},
+				[]Series{
+					newSeries(map[string]string{"a": "b"}, []Sample{
+						sample{3, 99}, sample{5, 99}, sample{6, 99}, sample{8, 99},
+						sample{9, 99}, sample{10, 99}, sample{13, 99}, sample{15, 99},
+						sample{16, 99}, sample{17, 99},
+					}),
+				},
+			},
+			exp: map[string][]sample{`{a="b"}`: {
+				sample{0, 0}, sample{1, 0}, sample{2, 0}, sample{3, 99},
+				sample{4, 0}, sample{5, 99}, sample{6, 99}, sample{7, 59},
+				sample{8, 59}, sample{9, 59}, sample{10, 59}, sample{11, 59},
+				sample{13, 99}, sample{15, 99}, sample{16, 99}, sample{17, 99},
+				sample{20, 0}, sample{22, 0},
+			}},
+		},
+	}
+
+	for _, c := range cases {
+		tmpdir, err := ioutil.TempDir("", "data")
+		testutil.Ok(t, err)
+
+		// Creating blocks.
+		for i, series := range c.blockSeries {
+			symbols := make(map[string]struct{})
+			for _, s := range series {
+				for _, l := range s.Labels() {
+					symbols[l.Name] = struct{}{}
+					symbols[l.Value] = struct{}{}
+				}
+			}
+			css := &MockChunkSeriesSet{
+				ss: newMockSeriesSet(series),
+			}
+			createNewBlock(t, tmpdir, c.mint[i], c.maxt[i], css, symbols)
```

Comment:
It would be difficult to control the time and the value of the sample to test the merging exactly. So I would prefer to keep it the way it is so that samples can be explicitly mentioned.

One solution is `createBlock` take optional samples.

Reply:
`One solution is createBlock take optional samples.`
yes this is what I meant.

Reply:
Wow! Refactoring `createBlock` to take series made me delete ~200 lines which I had to create block. I wish this was there before!

Reply:
Nice!